File: [23.12]Design-Pattern/Java/src/behavioral/memento/Main.java
package behavioral.memento;

import java.util.Stack;

public class Main {
    public static void main(String[] args) {
        Stack<Memento> mementos = new Stack<>();
        Adventurer adventurer = new Adventurer("전사", 10);

        System.out.println("초기 정보: " + adventurer.toString());

        mementos.push(adventurer.createMemento());
        adventurer.level = 12;
        adventurer.job = "법사";

        System.out.println("변경 후: " + adventurer.toString());

        adventurer.setInfo(mementos.pop());

        System.out.println("되돌리기: " + adventurer.toString());
    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/behavioral/memento/Main.kt
package behavioral.memento

import java.util.Stack

fun main() {
    val stack = Stack<Memento>()
    val adventurer = Adventurer("전사", 10)

    println("초기 값: $adventurer")

    stack.push(adventurer.createMemento())
    adventurer.job = "법사"
    adventurer.level = 12

    println("변경 후: $adventurer")

    adventurer.setData(stack.pop())
    println("되돌리기: $adventurer")
}